### DISCUSSION TITLE
Is there a way to make a single line of a text editor read-only?

### DISCUSSION THREAD

**Discussion by keith-gamble** [microsoft/vscode-discussions#282] [Extension Development QnA] [answered]
I am working to build an extension to enable editing of a different applications code, through VS.

The other application stores its python functions in an HTML safe encoded string, in a json file. So in order to edit it, I am unwrapping that content, decoding it, and providing a text document to the right for my user to edit the code in. 

### Example Application Code
```json
"script": "\tprint(\"Hello world\")"
```

I need the python document to be editable, but virtual, so I created a `FileSystemProvider` that stores the file in-memory to retain editing capabilities.

The problem with this, is that the `script` key is actually the contents of a function, but the function header is not present in the stored script. So I need to ad-hoc add it so that syntax highlighting works, and then ad-hoc remove it before I store it back into the json file again.

So I am adding the definition right before the code (in this example) with `def runAction(self, event):\n`

### What I am displaying when the user edits the script:
```py
def runAction(self, event):
	print("Hello world")
```

However I would ideally like to make the first line un-editable, that way I can still pull out the definition by replacing the string I added in the first place.

The reason I don't want to just cut out line 0 before I re-encode the script, is because some of the definitions also have docstrings so I will end up with multiple lines that I should ad-hoc add and remove.

Here is how I am currently opening the text document:
```js
function openIgnitionCode(documentUri, lineNumber, codeType) {
	
	const filePath = documentUri.path;

	// flint:transform.py?filePath=//myFolder/myFile.json&line=5
	const uri = vscode.Uri.parse(`${URI_SCHEME}:/flint/${codeType.fileName}.py?filePath=${filePath}&line=${lineNumber}&codeType=${codeType.codeKey}`);

	let code = getLineCode(filePath, lineNumber, codeType);

	// Decode the code
	code = decodeCodeText(code);

	// Insert the function definition
	code = insertFunctionDefinition(code, codeType);

	// Create a new document and open it in the editor

	FLINT_FS.writeFile(uri, Buffer.from(code), { create: true, overwrite: true });

	vscode.workspace.openTextDocument(uri).then(doc => {
		vscode.window.showTextDocument(doc, {preview: false, viewColumn: vscode.ViewColumn.Beside});
	});
}
```

I am open to suggestions, this is my first VS Code extension I plan to share with people, so thanks for any patience!

**Comment by isidorn** [comment, marked as answer]
I think it is not possible to make just one line of an editor read-only.
